### ISSUE TITLE
tooltip:  disconnected tooltips when parent element is detached/removed.

### ISSUE THREAD

**Issue by djfd** [angular/material#8033]
Hi, 
it is #8032 continuation

there is issue with md-tooltip design causing floating around alone widow lost tooltips

because tooltipParent is assigned to be documentBody then tooltip cannot detect its nesting element (actually parent) changes like to detach/removal from DOM.

We need either to change tooltipParent or have some way to detect parent element disappear/hide/remove and behave accordingly to remove tooltip
 
@ThomasBurleson
is it still general question need to be asked on forum? please read more carefully next time

**Comment by djfd** [comment]
eg can try next markup to have a proper parent
```
<div style="position:relative;">
    <div style="position:fixed; top:0;left:0;">
        <div class="md-content" ng-transclude></div>
    </div>
</div>
```

thus can catch parent's removal (eg. parent.remove()) using element.on('$destroy'), ...) and destroy self properly, also will be hidden if parent just detached, without actual removal

**Comment by topherfangio** [comment]
I believe I understand what you are saying, but can you please provide a Codepen which demonstrates the issue so that we have something to reference and debug when we look into this?

**Comment by djfd** [comment]
Hi topherfangio,

please check http://s.codepen.io/djFD/debug/qZoMoN

Does it work for you? for me it shows after clicking the button: see attachment
![iamlost](https://cloud.githubusercontent.com/assets/126338/14536280/eefe672a-02b4-11e6-8e63-f4c519030360.png)

**Comment by topherfangio** [comment]
@djfd - It appears that the button is still visible, and clickable, in that Codepen. After the first click, the tooltip is still shown because the button is still there. After the second click, the button disappears and so does the tooltip. Am I missing something? Also, would you mind providing the editable version of the Codepen so we can fork it and play with it? :smile:

**Comment by djfd** [comment]
well, possible I can. just do not know how to share it. any advise?

hmm, interesting, for me the button disappeared after first click, did you see any errors in console?

BTW, I am using Firefox (45.0.1), try it if possible

**Comment by topherfangio** [comment]
Thanks @djfd, I definitely see the issue and it's strange that on Firefox the button disappears immediately, but on Chrome the button disappears after the second click. We'll look into this, but I'm curious if you're actually doing `.detach()` in your code, or if you're using something else?

I'm curious because I don't think `.detach()` actually destroys the scope, and we should be checking for the scope destroy...

**Comment by djfd** [comment]
well, I do actually do detach. there is undo button that disappears if there is not any edits anymore... it is why...

yeah, detach does not fire $destroy on scope, neither fires $destroy on element, so possible it would be better to make remove instead.

in meanwhile I made my own version of it, wl-tooltip (watcher-less, to address the issue #8029 as well), now finishing debugging

**Comment by clshortfuse** [comment]
@topherfangio the Chrome click issue appears on angular 1.4.6

Angular 1.4.7 is fine. Here's a codepen fork with master and 1.4.7 

https://codepen.io/shortfuse/pen/eZrKOO

**Comment by djfd** [comment]
@clshortfuse
sorry, never did that so do not know how to proceed

**Comment by djfd** [comment]
Hi,

with a few minor CSS changes I did manage to attach tooltip to its own parent. That is parent removal (or detaching, or just CSS visibility changes) will also affect tooltip, resolving the issue. See modified example http://codepen.io/djFD/full/PNRLWq/ (wl-tooltip section)

There is the patch attached [wl-tooltip.js.diff.txt](https://github.com/angular/material/files/232790/wl-tooltip.js.diff.txt), the only difference with the master is an absence of 'md-visible'/'md-direction' attributes observer (just because I was not need them at the code writing time).

BTW, the following code (https://github.com/angular/material/blob/master/src/components/tooltip/tooltip.js#L110) looks wrong for me:
```
.forEach(function (mutation) {              
              if (mutation.attributeName === 'md-visible') {
                if (!scope.visibleWatcher)
                  scope.visibleWatcher = scope.$watch('visible', onVisibleChanged );
              }
```

I cannot get a point, why to turn on watcher if we already got a change inside attribute observer? Why just do not change the visibility accordingly, why to arm watcher?
